pkg/receive/handler.go

```diff
@@ -464,20 +465,57 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	}
 	h.mtx.RUnlock()
 
-	err := h.parallelizeRequests(ctx, tenant, replicas, wreqs)
-	if errs, ok := err.(terrors.MultiError); ok {
-		if uint64(countCause(errs, isNotReady)) >= (h.options.ReplicationFactor+1)/2 {
-			return tsdb.ErrNotReady
-		}
-		if uint64(countCause(errs, isConflict)) >= (h.options.ReplicationFactor+1)/2 {
-			return errors.Wrap(conflictErr, "did not meet replication threshold")
-		}
-		if uint64(len(errs)) >= (h.options.ReplicationFactor+1)/2 {
-			return errors.Wrap(err, "did not meet replication threshold")
+	ctx, cancel := context.WithTimeout(ctx, h.options.ForwardTimeout)
+	success := uint64(0)
+
+	n, ec := h.parallelizeRequests(ctx, tenant, replicas, wreqs)
+	defer func() {
+		go func() {
+			for {
+				err, more := <-ec
+				// Exhaust the channel, letting remaining unnecessary requests
+				// finish asnychronously.
+				if !more {
+					cancel()
+					return
+				}
+				level.Debug(h.logger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
+			}
+		}()
+	}()
+
+	var errs terrors.MultiError
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-ec:
+			if err == nil {
+				success++
+				if success >= (h.options.ReplicationFactor/2)+1 {
+					return nil
+				}
+			}
+			errs.Add(err)
+			n--
+			if n > 0 {
+				// Not done yet need to keep going.
+				continue
+			}
+
+			if uint64(countCause(errs, isNotReady)) >= (h.options.ReplicationFactor+1)/2 {
```

Comment:
Maybe worth to keep this important number in some function.. `(h.options.ReplicationFactor+1)/2 `